static/js/src/advantage/subscribe/react/PurchaseModal.jsx

```diff
@@ -0,0 +1,319 @@
+        )
+      ) {
+        setError(
+          <>
+            We were unable to verify your credit card. Check the details and try
+            again. Contact{" "}
+            <a href="https://ubuntu.com/contact-us">Canonical sales</a> if the
+            problem persists.
+          </>
+        );
+      } else {
+        setError(
+          <>
+            We were unable to process the payment. Check the details and try
+            again. Contact{" "}
+            <a href="https://ubuntu.com/contact-us">Canonical sales</a> if the
+            problem persists.
+          </>
+        );
+      }
+      setStep(1);
+    }
+  }, [purchaseError]);
+
+  useEffect(() => {
+    if (pendingPurchase?.status === "done") {
+      // Remove the product selection from the local storage
+      localStorage.removeItem("ua-subscribe-state");
+
+      //redirect
+      if (window.isGuest) {
+        location.href = `/advantage/subscribe/thank-you?email=${encodeURIComponent(
+          pendingPurchase?.invoice?.customerEmail
+        )}`;
+      } else {
+        location.pathname = "/advantage";
+      }
+    }
+  }, [pendingPurchase]);
+
+  const onSubmit = (values, actions) => {
+    setError(null);
+    paymentMethodMutation.mutate(values, {
+      onSuccess: (data, variables) => {
+        window.accountId = data.accountId;
+        setStep(2);
+        queryClient.setQueryData(
+          "userInfo",
+          getUserInfoFromVariables(data, variables)
+        );
+
+        actions.setSubmitting(false);
+      },
+      onError: (error) => {
+        if (error.message === "email_already_exists") {
+          //Email already exists
+          setError(
+            <>
+              An Ubuntu One account with this email address exists. Please{" "}
+              <a href="/login">sign in</a> to your account first.
+            </>
+          );
+        } else {
+          // Tries to match the error with a known error code and defaults to a generic error if it fails
+          setError(
+            getErrorMessage({ message: "", code: error.message }) ?? (
+              <>
+                Sorry, there was an unknown error with your credit card. Check
+                the details and try again. Contact{" "}
+                <a href="https://ubuntu.com/contact-us">Canonical sales</a> if
+                the problem persists.
+              </>
+            )
+          );
+        }
+
+        actions.setSubmitting(false);
+      },
+    });
+  };
+
+  const onPayClick = () => {
+    purchaseMutation.mutate("", {
+      onSuccess: (data) => {
+        //start polling
+        setPendingPurchaseID(data);
+      },
+      onError: (error) => {
+        // An error happened
+        if (error.message.includes("can only make one purchase at a time")) {
+          setError(
+            <>
+              You already have a pending purchase. Please go to{" "}
+              <a href="/advantage/payment-methods">payment methods</a> to retry.
```

Comment:
Do we display pending purchases on the payment-methods page?

Reply:
So, my reasoning here was: If the error comes from invalid form data and can be retried, we display it above the form to let you retry. But if the error has to do with your account state or anything else that is out of your control then there is no point in showing you the form again since re-filling it would do nothing.
@wgx  what do you think?

Reply:
@ClementChaumel @anthonydillon Yep, this is fine. Payment methods is the only way to fix a "pending purchases" state - so this is good.